In [1]:
#importing required libraries
import pandas as pd
import numpy as np

In [2]:
#Reading the csv file
df=pd.read_csv(r"C:\Users\RAGHAVENDRA KUMAR\Downloads\2732079-Datasets (1)\Datasets\ethnic_diversity_data_pandas_accessing_task.csv")

In [3]:
df

Employee_Name       EmpID                  Position State   Zip  \
0                Brown, Mia  1103024456              Accountant I    MA  1450   
1      LaRotonda, William    1106026572              Accountant I    MA  1460   
2          Steans, Tyrone    1302053333              Accountant I    MA  2703   
3           Howard, Estelle  1211050782  Administrative Assistant    MA  2170   
4               Singh, Nan   1307059817  Administrative Assistant    MA  2330   
..                      ...         ...                       ...   ...   ...   
305          Daniele, Ann    1411071312      Sr. Network Engineer    CT  6033   
306          Lajiri,  Jyoti  1108028108      Sr. Network Engineer    MA  2169   
307  Semizoglou, Jeremiah     904013591      Sr. Network Engineer    MA  1545   
308              South, Joe  1308060959      Sr. Network Engineer    CT  6278   
309         Warfield, Sarah  1301052347      Sr. Network Engineer    MA  2453   

    Sex MaritalDesc CitizenDesc      EmploymentStatus     Department  \
0     F     Married  US Citizen                Active  Admin Offices   
1    M     Divorced  US Citizen                Active  Admin Offices   
2    M       Single  US Citizen                Active  Admin Offices   
3     F     Married  US Citizen  Terminated for Cause  Admin Offices   
4     F      Single  US Citizen                Active  Admin Offices   
..   ..         ...         ...                   ...            ...   
305   F     Married  US Citizen      Leave of Absence          IT/IS   
306  M      Married  US Citizen      Leave of Absence          IT/IS   
307  M      Married  US Citizen          Future Start          IT/IS   
308  M       Single  US Citizen                Active          IT/IS   
309   F     Widowed  US Citizen                Active          IT/IS   

     Salaries  age                       Race  
0      674.28   28  Black or African American  
1      674.28   50  Black or African American  
2      674.28   43                      White  
3    47822.00   44                      White  
4    61089.35   33                      White  
..        ...  ...                        ...  
305  37619.01   38                      White  
306  27603.00   34                      White  
307  27587.46   24                      Asian  
308  12126.53   31                      White  
309  37257.00   46                      Asian  

[310 rows x 13 columns]

In [4]:
#Info about dataset
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 310 entries, 0 to 309
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Employee_Name     310 non-null    object 
 1   EmpID             310 non-null    int64  
 2   Position          310 non-null    object 
 3   State             310 non-null    object 
 4   Zip               310 non-null    int64  
 5   Sex               310 non-null    object 
 6   MaritalDesc       310 non-null    object 
 7   CitizenDesc       310 non-null    object 
 8   EmploymentStatus  310 non-null    object 
 9   Department        310 non-null    object 
 10  Salaries          310 non-null    float64
 11  age               310 non-null    int64  
 12  Race              310 non-null    object 
dtypes: float64(1), int64(3), object(9)
memory usage: 31.6+ KB


In [5]:
#Statistical information about dataframe(df)
df.describe()

EmpID           Zip       Salaries         age
count  3.100000e+02    310.000000     310.000000  310.000000
mean   1.199745e+09   6569.732258   36670.102742   38.696774
std    1.829600e+08  16933.864054   21075.942369    9.258163
min    6.020003e+08   1013.000000       0.000000   23.000000
25%    1.101024e+09   1901.250000   23092.950000   31.000000
50%    1.203032e+09   2132.000000   34554.745000   39.000000
75%    1.378814e+09   2357.000000   51452.895000   46.000000
max    1.988300e+09  98052.000000  108304.000000   54.000000

# 1. What is the minimum salary in the dataset?

In [6]:
df["Salaries"].min()

0.0

In [7]:
df.loc[(df["Salaries"]==df["Salaries"].min())]

Employee_Name       EmpID                 Position State   Zip Sex  \
209  Anderson, Carol   1001417624  Production Technician I    MA  2169   F   

    MaritalDesc CitizenDesc        EmploymentStatus         Department  \
209    Divorced  US Citizen  Voluntarily Terminated  Production          

     Salaries  age   Race  
209       0.0   34  White

In [45]:
print("The minimum salary:",df["Salaries"].min())

The minimum salary: 0.0


# 2. How many employers are there without marriage i.e. Single?

In [8]:
df.loc[(df["MaritalDesc"]=="Single"),["Employee_Name","MaritalDesc"]]

Employee_Name MaritalDesc
2     Steans, Tyrone        Single
4          Singh, Nan       Single
7    Carter, Michelle       Single
8    Dietrich, Jenna        Single
10     Friedman, Gerry      Single
..                 ...         ...
298   Sweetwater, Alex      Single
302   Carr, Claudia  N      Single
303    Favis, Donald        Single
304   Roehrich, Bianca      Single
308         South, Joe      Single

[137 rows x 2 columns]

# Insights:
There are 137 single employees i.e, without married. 

# 3.What is the average salary of employees?

In [9]:
df["Salaries"].mean()

36670.10274193547

In [10]:
print("The average salary of employees :",df["Salaries"].mean())

The average salary of employees : 36670.10274193547


# 4. What are the different positions in dataset?

In [11]:
print("The different positions in dataset are :")
df["Position"].unique()

The different positions in dataset are :


array(['Accountant I', 'Administrative Assistant', 'Area Sales Manager',
       'BI Developer', 'BI Director', 'CIO', 'Data Architect',
       'Database Administrator', 'Data Analyst', 'Data Analyst ',
       'Director of Operations', 'Director of Sales', 'IT Director',
       'IT Manager - DB', 'IT Manager - Infra', 'IT Manager - Support',
       'IT Support', 'Network Engineer', 'President & CEO',
       'Production Manager', 'Production Technician I',
       'Production Technician II', 'Sales Manager', 'Senior BI Developer',
       'Shared Services Manager', 'Software Engineer',
       'Software Engineering Manager', 'Sr. Accountant', 'Sr. DBA',
       'Enterprise Architect', 'Principal Data Architect',
       'Sr. Network Engineer'], dtype=object)

# 5. Get all the details of employee who is getting highest salary?

In [12]:
df.loc[(df["Salaries"]==df["Salaries"].max())]

Employee_Name      EmpID                 Position State   Zip Sex  \
216  Kinsella, Kathleen    710007401  Production Technician I    MA  2170   F   

    MaritalDesc CitizenDesc        EmploymentStatus         Department  \
216     Married  US Citizen  Voluntarily Terminated  Production          

     Salaries  age   Race  
216  108304.0   27  White

# 6.Which 'Department' has more employees?

In [112]:
df["Department"].value_counts()

Production              208
IT/IS                    50
Sales                    31
Admin Offices            10
Software Engineering     10
Executive Office          1
Name: Department, dtype: int64

In [13]:
df2=df.groupby("Department")

In [14]:
df2.ngroups

6

In [21]:
df2.count()

Employee_Name  EmpID  Position  State  Zip  Sex  \
Department                                                              
Admin Offices                    10     10        10     10   10   10   
Executive Office                  1      1         1      1    1    1   
IT/IS                            50     50        50     50   50   50   
Production                      208    208       208    208  208  208   
Sales                            31     31        31     31   31   31   
Software Engineering             10     10        10     10   10   10   

                      MaritalDesc  CitizenDesc  EmploymentStatus  Salaries  \
Department                                                                   
Admin Offices                  10           10                10        10   
Executive Office                1            1                 1         1   
IT/IS                          50           50                50        50   
Production                    208          208               208       208   
Sales                          31           31                31        31   
Software Engineering           10           10                10        10   

                      age  Race  
Department                       
Admin Offices          10    10  
Executive Office        1     1  
IT/IS                  50    50  
Production            208   208  
Sales                  31    31  
Software Engineering   10    10

In [22]:
df2.groups

{'Admin Offices': [0, 1, 2, 3, 4, 5, 287, 288, 299, 300], 'Executive Office': [73], 'IT/IS': [33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67, 68, 69, 70, 71, 72, 284, 285, 286, 301, 302, 303, 304, 305, 306, 307, 308, 309], 'Production       ': [53, 74, 75, 76, 77, 78, 79, 80, 81, 82, 83, 84, 85, 86, 87, 88, 89, 90, 91, 92, 93, 94, 95, 96, 97, 98, 99, 100, 101, 102, 103, 104, 105, 106, 107, 108, 109, 110, 111, 112, 113, 114, 115, 116, 117, 118, 119, 120, 121, 122, 123, 124, 125, 126, 127, 128, 129, 130, 131, 132, 133, 134, 135, 136, 137, 138, 139, 140, 141, 142, 143, 144, 145, 146, 147, 148, 149, 150, 151, 152, 153, 154, 155, 156, 157, 158, 159, 160, 161, 162, 163, 164, 165, 166, 167, 168, 169, 170, 171, 172, ...], 'Sales': [6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 54, 281, 282, 283], 'Software Engineering': [289, 290, 291, 292, 293, 294, 295

# Insights:
Production department has more employees i.e, 208 when compared to other departments.

# 7. How many 'female' employees are working in the company?

In [46]:
df.loc[(df["Sex"]=="F"),["Employee_Name","Sex"]]

Employee_Name Sex
0           Brown, Mia   F
3      Howard, Estelle   F
4          Singh, Nan    F
5     Smith, Leigh Ann   F
6     Bunbury, Jessica   F
..                 ...  ..
300  Boutwell, Bonalyn   F
302   Carr, Claudia  N   F
304   Roehrich, Bianca   F
305     Daniele, Ann     F
309    Warfield, Sarah   F

[177 rows x 2 columns]

# Insights:
There are 177 female employees working in the company.

# 8. Who are the 'emplDs' 'Non-Citizen' of US (.'Citizen Desc' column)

In [23]:
df.loc[(df["CitizenDesc"]=="Non-Citizen"),["Employee_Name","EmpID","CitizenDesc"]]

Employee_Name       EmpID  CitizenDesc
165  Cierpiszewski, Caroline    1012023295  Non-Citizen
222         Tavares, Desiree    1405067501  Non-Citizen
271     Homberger, Adrienne  J  1207046956  Non-Citizen
297               True, Edward  1102024057  Non-Citizen

# 9. What is the average salary of employees who are Asian (refer: 'Race' column)

In [27]:
df3=df.groupby("Race")
df3.ngroups

6

In [42]:
df3["Salaries"].mean()

Race
American Indian or Alaska Native    14287.082500
Asian                               40662.373235
Black or African American           36293.186140
Hispanic                            27110.425000
Two or more races                   32971.810000
White                               37085.061865
Name: Salaries, dtype: float64

In [39]:
df4=df3.get_group("Asian")
df4

Employee_Name       EmpID                  Position State   Zip  \
5          Smith, Leigh Ann   711007713  Administrative Assistant    MA  1844   
19         Villanueva, Noah  1111030503        Area Sales Manager    ME  4063   
44         Johnson, Noelle   1003018246    Database Administrator    MA  2301   
47        Petrowsky, Thelma  1108027853              Data Analyst    MA  1886   
75           Bozzi, Charles  1303054580        Production Manager    MA  1901   
80        Immediato, Walter  1403065874        Production Manager    MA  2128   
110        Kretschmer, John  1311062610   Production Technician I    MA  1801   
112          Ngodup, Shari   1403066020   Production Technician I    MA  1810   
117      Ybarra, Catherine   1001268402   Production Technician I    MA  2458   
132           Zima, Colleen  1211051232   Production Technician I    MA  1730   
139        Harrell, Ludwick  1406068241   Production Technician I    MA  1776   
146        Sutwell, Barbara  1407069061   Production Technician I    MA  2718   
161            Stoica, Rick  1106026896   Production Technician I    MA  2301   
164         Biden, Lowan  M  1412071844   Production Technician I    MA  2027   
178         Barbara, Thomas  1312063507   Production Technician I    MA  2062   
182         Garneau, Hamish  1104025414   Production Technician I    MA  1550   
183           Goyal, Roxana  1312063675   Production Technician I    MA  1864   
185           Leach, Dallas  1408069409   Production Technician I    MA  1810   
188          Power, Morissa  1102024274   Production Technician I    MA  1742   
210           Becker, Scott  1110029777   Production Technician I    MA  2452   
220         Rivera, Haley    1405067642   Production Technician I    MA  2171   
221          Sewkumar, Nori   807010161   Production Technician I    MA  2191   
228       Pelletier, Ermine  1202031821  Production Technician II    MA  2472   
234      Petingill, Shana    1103024843  Production Technician II    MA  2050   
238       Langford, Lindsey  1010022030  Production Technician II    MA  2090   
279         Oliver, Brooke   1001856521  Production Technician II    MA  2492   
284                Le, Binh  1009919930       Senior BI Developer    MA  1886   
285           Navathe, Kurt  1009919960       Senior BI Developer    MA  2056   
286           Wang, Charlie  1009919970       Senior BI Developer    MA  1887   
293          Martin, Sandra  1303054625         Software Engineer    MA  2135   
294         Patronick, Luke  1112030979         Software Engineer    MA  1844   
300       Boutwell, Bonalyn  1307060188            Sr. Accountant    MA  2468   
307  Semizoglou, Jeremiah     904013591      Sr. Network Engineer    MA  1545   
309         Warfield, Sarah  1301052347      Sr. Network Engineer    MA  2453   

    Sex MaritalDesc          CitizenDesc        EmploymentStatus  \
5     F     Married           US Citizen  Voluntarily Terminated   
19   M       Single           US Citizen                  Active   
44    F     Married           US Citizen        Leave of Absence   
47    F     Married           US Citizen                  Active   
75   M       Single           US Citizen  Voluntarily Terminated   
80   M      Married           US Citizen  Voluntarily Terminated   
110  M      Married           US Citizen                  Active   
112   F   Separated           US Citizen                  Active   
117   F      Single           US Citizen  Voluntarily Terminated   
132   F     Widowed           US Citizen                  Active   
139  M      Married           US Citizen                  Active   
146   F      Single  Eligible NonCitizen                  Active   
161  M      Married           US Citizen                  Active   
164   F    Divorced           US Citizen                  Active   
178  M      Married           US Citizen  Voluntarily Terminated   
182  M       Single           US Citizen                  Active   
183   F     Married   

In [40]:
df4["Salaries"].mean()

40662.3732352941

In [41]:
print("The average salary of employees who are Asian:",df4["Salaries"].mean())

The average salary of employees who are Asian: 40662.3732352941


# 10. Show all the employs who is not currently working in the company (check Employmentstatus` column)?

In [43]:
df["EmploymentStatus"].unique()

array(['Active', 'Terminated for Cause', 'Voluntarily Terminated',
       'Future Start', 'Leave of Absence'], dtype=object)

In [44]:
df.loc[(df["EmploymentStatus"]=='Terminated for Cause') | (df["EmploymentStatus"]=="Voluntarily Terminated"),["Employee_Name"]]

Employee_Name
3             Howard, Estelle
5            Smith, Leigh Ann
6            Bunbury, Jessica
11            Gill, Whitney  
13            Guilianno, Mike
..                        ...
294           Patronick, Luke
297              True, Edward
301  Ait Sidi, Karthikeyan   
303           Favis, Donald  
304          Roehrich, Bianca

[103 rows x 1 columns]

# Insights:
There are 103 employees who are not working in the company(they are Voluntarily Terminated or Terminated for Cause)

# Q2. In the attached iris dataset, do all slicing, indexing and Filtering of columns with various conditions.

In [47]:
#Reading the iris_dataset using read_csv
d1=pd.read_csv(r"C:\Users\RAGHAVENDRA KUMAR\Downloads\2732079-Datasets (1)\Datasets\iris.csv")

In [57]:
d2=d1.copy()

In [58]:
#Removing Unnamed:0 column 
d2=d2.drop(columns="Unnamed: 0")

In [59]:
d2

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)  \
0                  5.1               3.5                1.4               0.2   
1                  4.9               3.0                1.4               0.2   
2                  4.7               3.2                1.3               0.2   
3                  4.6               3.1                1.5               0.2   
4                  5.0               3.6                1.4               0.2   
..                 ...               ...                ...               ...   
145                6.7               3.0                5.2               2.3   
146                6.3               2.5                5.0               1.9   
147                6.5               3.0                5.2               2.0   
148                6.2               3.4                5.4               2.3   
149                5.9               3.0                5.1               1.8   

     target target_names  
0         0       setosa  
1         0       setosa  
2         0       setosa  
3         0       setosa  
4         0       setosa  
..      ...          ...  
145       2    virginica  
146       2    virginica  
147       2    virginica  
148       2    virginica  
149       2    virginica  

[150 rows x 6 columns]

# Indexing on Rows using loc[]

In [62]:
d2.loc[[0]]

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)  \
0                5.1               3.5                1.4               0.2   

   target target_names  
0       0       setosa

In [64]:
d2.loc[[0,1,3]]

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)  \
0                5.1               3.5                1.4               0.2   
1                4.9               3.0                1.4               0.2   
3                4.6               3.1                1.5               0.2   

   target target_names  
0       0       setosa  
1       0       setosa  
3       0       setosa

# Slicing on rows using loc[] and iloc[]

In [65]:
d2.loc[0:2]

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)  \
0                5.1               3.5                1.4               0.2   
1                4.9               3.0                1.4               0.2   
2                4.7               3.2                1.3               0.2   

   target target_names  
0       0       setosa  
1       0       setosa  
2       0       setosa

In [66]:
d2.iloc[0:2]

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)  \
0                5.1               3.5                1.4               0.2   
1                4.9               3.0                1.4               0.2   

   target target_names  
0       0       setosa  
1       0       setosa

In [67]:
d2.iloc[0::2]

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)  \
0                  5.1               3.5                1.4               0.2   
2                  4.7               3.2                1.3               0.2   
4                  5.0               3.6                1.4               0.2   
6                  4.6               3.4                1.4               0.3   
8                  4.4               2.9                1.4               0.2   
..                 ...               ...                ...               ...   
140                6.7               3.1                5.6               2.4   
142                5.8               2.7                5.1               1.9   
144                6.7               3.3                5.7               2.5   
146                6.3               2.5                5.0               1.9   
148                6.2               3.4                5.4               2.3   

     target target_names  
0         0       setosa  
2         0       setosa  
4         0       setosa  
6         0       setosa  
8         0       setosa  
..      ...          ...  
140       2    virginica  
142       2    virginica  
144       2    virginica  
146       2    virginica  
148       2    virginica  

[75 rows x 6 columns]

# Slicing on Rows and columns

In [70]:
d2.iloc[0:3,1:]

sepal width (cm)  petal length (cm)  petal width (cm)  target target_names
0               3.5                1.4               0.2       0       setosa
1               3.0                1.4               0.2       0       setosa
2               3.2                1.3               0.2       0       setosa

In [84]:
d2.loc[0:3,"sepal length (cm)":"petal length (cm)"]

sepal length (cm)  sepal width (cm)  petal length (cm)
0                5.1               3.5                1.4
1                4.9               3.0                1.4
2                4.7               3.2                1.3
3                4.6               3.1                1.5

In [85]:
d2.iloc[0:3,0:3]

sepal length (cm)  sepal width (cm)  petal length (cm)
0                5.1               3.5                1.4
1                4.9               3.0                1.4
2                4.7               3.2                1.3

# Filtering of columns based on different conditions

# What are the target_names whose sepallength<4.6 and petal length>1.5 

In [89]:
d2.loc[(d2["sepal length (cm)"]>4.6) & (d2["petal length (cm)"]>1.5),["sepal length (cm)","petal length (cm)","target_names"]]

sepal length (cm)  petal length (cm) target_names
5                  5.4                1.7       setosa
11                 4.8                1.6       setosa
18                 5.7                1.7       setosa
20                 5.4                1.7       setosa
23                 5.1                1.7       setosa
..                 ...                ...          ...
145                6.7                5.2    virginica
146                6.3                5.0    virginica
147                6.5                5.2    virginica
148                6.2                5.4    virginica
149                5.9                5.1    virginica

[113 rows x 3 columns]

# Filter the dataframe  whose sepallength<4.9 and petal length>1.5 belongs to setosa

In [93]:
d2.loc[(d2["sepal length (cm)"]<4.9) & (d2["petal length (cm)"]>1.5) & (d2["target_names"]=="setosa"),["sepal length (cm)","petal length (cm)","target_names"]]

sepal length (cm)  petal length (cm) target_names
11                4.8                1.6       setosa
24                4.8                1.9       setosa
29                4.7                1.6       setosa
30                4.8                1.6       setosa

# Filter the dataframe  whose sepallength<7.5 and petal length>6.2 belongs to virginica

In [100]:
d2.loc[(d2["sepal length (cm)"]<7.5) & (d2["petal length (cm)"]>6.2) & (d2["target_names"]=="virginica"),["sepal length (cm)","petal length (cm)","target_names"]]

sepal length (cm)  petal length (cm) target_names
107                7.3                6.3    virginica

# Filter the dataframe  whose sepallength<7and petal length>5 belongs to versicolor

In [110]:
d2.loc[(d2["sepal length (cm)"]<7) & (d2["petal length (cm)"]>5) & (d2["target_names"]=="versicolor"),["sepal length (cm)","petal length (cm)","target_names"]]

sepal length (cm)  petal length (cm) target_names
83                6.0                5.1   versicolor